captum/optim/_utils/atlas.py

```diff
@@ -0,0 +1,250 @@
+from typing import Callable, List, Tuple, Union, cast
+
+import torch
+
+
+def normalize_grid(
+    xy_grid: torch.Tensor,
+    min_percentile: float = 0.01,
+    max_percentile: float = 0.99,
+    relative_margin: float = 0.1,
+) -> torch.Tensor:
+    """
+    Remove outliers from an xy coordinate grid tensor, and rescale it to [0, 1].
+
+    Args:
+
+        xy_grid (torch.tensor): The xy coordinate grid tensor to normalize,
+            with a shape of: [n_points, n_axes].
+        min_percentile (float, optional): The minimum percentile to use when
+            normalizing the tensor. Value must be in the range [0, 1].
+        max_percentile (float, optional): The maximum percentile to use when
+            normalizing the tensor. Value must be in the range [0, 1].
+        relative_margin (float, optional): The relative margin to use when
+            normalizing the tensor.
+
+    Returns:
+        normalized_grid (torch.tensor): A normalized xy coordinate grid tensor.
+    """
+
+    assert xy_grid.dim() == 2
+    assert 0.0 <= min_percentile <= 1.0
+    assert 0.0 <= max_percentile <= 1.0
+
+    mins = torch.quantile(xy_grid, min_percentile, dim=0)
+    maxs = torch.quantile(xy_grid, max_percentile, dim=0)
+
+    mins = mins - relative_margin * (maxs - mins)
+    maxs = maxs + relative_margin * (maxs - mins)
+
+    normalized_grid = torch.max(torch.min(xy_grid, maxs), mins)
+    normalized_grid = normalized_grid - normalized_grid.min(0)[0]
+    return normalized_grid / normalized_grid.max(0)[0]
+
+
+def calc_grid_indices(
+    xy_grid: torch.Tensor,
+    grid_size: Tuple[int, int],
+    x_extent: Tuple[float, float] = (0.0, 1.0),
+    y_extent: Tuple[float, float] = (0.0, 1.0),
+) -> List[List[torch.Tensor]]:
+    """
+    Create sets of grid cell indices of a specified size for an irregular grid.
+
+    Args:
+
+        xy_grid (torch.tensor): The xy coordinate grid activation samples, with a shape
+            of: [n_points, 2].
+        grid_size (Tuple[int, int]): The grid_size of grid cells to use. The grid_size
+            variable should be in the format of: [width, height].
+        x_extent (Tuple[float, float], optional): The x axis range to use.
+        y_extent (Tuple[float, float], optional): The y axis range to use.
+
+    Returns:
+        indices (list of list of tensor): Grid cell indices for the irregular grid.

```

Comment:
@ProGamerGov, I think that the following is a good description of `indices`. Do you mind adding that documentation as a comment starting from line 64 ?
```
The calc_grid_indices function draws a 2D grid across the irregular grid of points, and then groups point indices based on the grid cell they fall within. These grid cells are then denoted by their x and y axis positions. The grid cells are then filled with 1D tensors that have anywhere from 0 to n_indices values in them.

The grid cell have their indices stored in a list of list of tensors. The outer lists represent the x axis cell coordinates, while the inner lists the y axis cell coordinates.

Below is an example of the indice list format for a grid size of (3, 3):

indices = [ x1[y1, y2, y3], x2[y1, y2, y3], x3[y1, y2, y3], ] 
 Each pair of x and y has a tensor with a size of 0 to n_indices
```

Reply:
Regarding `extract_grid_vectors` function, you can describe high level idea. No need to describe each variable. With the description I meant describing high level idea and expected dimensionality for the matrices or vectors that we return and make available in the public API.

Reply:
@NarineK Awesome! I think that I've summarized it nicely with the following:

```
def calc_grid_indices(
    xy_grid: torch.Tensor,
    grid_size: Tuple[int, int],
    x_extent: Tuple[float, float] = (0.0, 1.0),
    y_extent: Tuple[float, float] = (0.0, 1.0),
) -> List[List[torch.Tensor]]:
    """
    This function draws a 2D grid across the irregular grid of points, and then groups
    point indices based on the grid cell they fall within. The grid cells are then
    filled with 1D tensors that have anywhere from 0 to n_indices values in them. The
    sets of grid indices can then be used with the extract_grid_vectors function to 
    create atlas grid cell direction vectors.

    Indices are stored for grid cells in an xy matrix, where the outer lists represent
    x positions and the inner lists represent y positions. Each grid cell is filled
    with 1D tensors that have anywhere from 0 to n_indices index values inside them. 

    Below is an example of the index list format for a grid_size of (3, 3):
    indices = [x1[y1, y2, y3], x2[y1, y2, y3], x3[y1, y2, y3]]

    Grid cells would then be ordered like this, where each cell contains a list of
    indices for that particular cell:
    indices = [[0, 1, 2], [3, 4, 5], [6, 7, 8]]

    Args:
        xy_grid (torch.tensor): The xy coordinate grid activation samples, with a shape
            of: [n_points, 2].
        grid_size (Tuple[int, int]): The grid_size of grid cells to use. The grid_size
            variable should be in the format of: [width, height].
        x_extent (Tuple[float, float], optional): The x axis range to use.
            Default: (0.0, 1.0)
        y_extent (Tuple[float, float], optional): The y axis range to use.
            Default: (0.0, 1.0)
    Returns:
        indices (list of list of torch.Tensors): List of lists of grid indices
            stored inside tensors to use. Each 1D tensor of indices has a size of:
            0 to n_indices.
    """
```

Reply:
thank you for adding the documentation, @ProGamerGov! Since the returned tensors represent `List[List[torch.Tensor]]` wouldn't then the output look something like this ?

```indices = [[tensor([0, 5]), tensor([1]), tensor([2,3])], [tensor([]), tensor([4]), tensor([])], [tensor([6,7,8]), tensor([]), tensor([])]]```

instead of `[[0, 1, 2], [3, 4, 5], [6, 7, 8]]` ?

Reply:
@ProGamerGov, overall the PR looks good. Thank you for addressing my comments. I was wondering if my understanding here is valid and whether we can update the doc accordingly.

Reply:
@NarineK Oops, sorry I missed this comment! So for the example, I should have probably mentioned that `[[0, 1, 2], [3, 4, 5], [6, 7, 8]] ` is describing the order for the 3x3 grid example. I've updated the doc to explain this:

```
    """
    Below is an example of the index list format for a grid_size of (3, 3):
    indices = [x1[y1, y2, y3], x2[y1, y2, y3], x3[y1, y2, y3]]

    Grid cells would then be ordered like this:
    indices = [[0, 1, 2], [3, 4, 5], [6, 7, 8]]

    Each cell in the above example would contain a list of indices inside a tensor for
    that particular cell, like this:
    indices = [
        [tensor([0, 5]), tensor([1]), tensor([2, 3])],
        [tensor([]), tensor([4]), tensor([])],
        [tensor([6, 7, 8]), tensor([]), tensor([])],
    ]

    Args:
    """
```